exporter: Enable to specify the compression type for all layers of the finally exported image

ktock opened the pull request:
resolves: https://github.com/moby/buildkit/issues/1911

Currently, exporter's `compression` flag is applied only to newly created layers. So the exported image can be a mix of several compression types (uncompressed and gzip) even if the `compression` type is specified.

It would be great if we can specify the compression type of all layers of the finally exported image.

This commit introduces a new exporter flag `force-compression` enabling to specify the compression type of the exported layers. If layers don't match the type specified by `compression` option, they are forcefully converted to the targeting type.

In addition to making sure the final compression type, this flag, in the future, will also help users to quickly catch up and try with updates & trends on novel image layer formats (e.g. zstandard, estargz, zstd:chunked, etc).

### How compression variant contents are tracked?

When a user specifies a compression type different from the blob's original (i.e. pulled) one, a conversion happens. Once the blob is converted, the result contents are cached to the local content store. `*immutableRef` tracks its compression variant contents using its lease and the following label to the original (pulled) content blob.

- `buildkit.io/compression/digest.<compression type>` = `<sha256 digest>`

During converting compression type, this label is looked up first. If there is no targeting label or the content is lost in the content store, conversion happens.

### Test of resolving #1911

```bash
BUILD="buildctl build --frontend=dockerfile.v0 --local context=/tmp/tmpimg --local dockerfile=/tmp/tmpimg "
$BUILD --output type=oci,dest=/tmp/uncompressed.tar,oci-mediatypes=true,compression=uncompressed
$BUILD --output type=oci,dest=/tmp/gzip.tar,oci-mediatypes=true,compression=gzip,force-compression=true
```

The above commands produce `gzip.tar` with gzip-compressed layers:
```json
{
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "schemaVersion": 2,
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "digest": "sha256:b105e3a8ccf94db2a87393ad4bdb1bd938750f1e8a72b54f4ff15573afdf7396",
    "size": 1267
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
      "digest": "sha256:9d48c3bd43c520dc2784e868a780e976b207cbf493eaff8c6596eb871cbd9609",
      "size": 2789669
    },
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
      "digest": "sha256:49f20b6e614f787092825b2e2f22ad62be1c5e6fb9aec85276039383df91c6a7",
      "size": 103
    },
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
      "digest": "sha256:3de9369605bb1c6f39b1559feb19f77f4aad8f5d0b4cca92a6ef827eb423a6fa",
      "size": 106
    },
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
      "digest": "sha256:99cd2a06d8a234d71e8c5dccafd0bdf46497ed6b2d5603f538d7608428844cd6",
      "size": 106
    }
  ]
}
```


@AkihiroSuda @tonistiigi @fuweid

AkihiroSuda commented on README.md:
```diff
@@ -230,8 +230,8 @@ Keys supported by image output:
-* `compression=[uncompressed,gzip]`: choose compression type for layer, gzip is default value
-
+* `compression=[uncompressed,gzip]`: choose compression type for layers newly created and cached, gzip is default value
+* `force-exporting-compression=[uncompressed,gzip]`: choose compression type for all layers (including already existing layers) for exporting. compression type specified by `compression` flag is respected by default.
```
Option name is too long. I suggest `compression-force` or `compression-all`.

AkihiroSuda replied:
Or can we replace the old `compression` opt? Was it ever useful?

ktock replied:
Renamed to `compression-all`.

AkihiroSuda commented on exporter/containerimage/writer.go:
```diff
@@ -177,6 +185,162 @@ func (ic *ImageWriter) exportLayers(ctx context.Context, compressionType compres
 	return out, nil
 }
 
+func (ic *ImageWriter) ensureCompressionType(ctx context.Context, remote *solver.Remote, forceCompressionType compression.Type) (*solver.Remote, error) {
```
Can we use containerd converter API as in `ctr-remote image convert` and `nerdctl image convert`

ktock replied:
Updated to use converter API.

ktock commented:
PTAL

AkihiroSuda commented on exporter/containerimage/writer.go:
```diff
@@ -177,6 +199,37 @@ func (ic *ImageWriter) exportLayers(ctx context.Context, compressionType compres
+	case compression.Uncompressed:
+		layerConvertFunc = uncompress.LayerConvertFunc
+	case compression.Gzip:
+		layerConvertFunc = gzipLayerConvertFunc
```
Can we move this converter func to the main repo later?

ktock replied:
Yes, I will that.

ktock commented:
@tonistiigi Thanks for taking look at this. 
How is this going?

ktock commented:
Can we move this forward?

tonistiigi reviewed (commented):
I don't understand how this does tracking of the state of the created blobs. Afaics they are not tracked by the cache manager and I don't see any leases around writers. iiuc the tracked blob is still created with the old code, ignoring compression settings and then that blob may be converted in a second pass.

As mentioned in https://github.com/moby/buildkit/issues/1911#issuecomment-748685650 the hardest part about this should be how to associate multiple blobs with a single record.

tonistiigi commented on exporter/containerimage/writer.go:
```diff
@@ -177,6 +199,37 @@ func (ic *ImageWriter) exportLayers(ctx context.Context, compressionType compres
+	}
+
+	// unlazy layers as converter uses layer contents in the content store
+	// TODO(ktock): un-lazy only layers whose type is different from the target, selectively.
```
It's also a sign that blob creation is not happening in the correct place as `Remote` already contains a reference to a blob(s).

ktock commented:
@tonistiigi Thanks for the review. Fixed the design to track compression variant contents as the following.

When a user specifies a compression type different from the blob's original (i.e. pulled) one, a conversion happens. Once the blob is converted, the result contents are cached to the local content store. `*immutableRef` tracks its compression variant contents using its lease and the following label of the original (pulled) content blob.

- `buildkit.io/compression/digest.<compression type>` = `<sha256 digest>`

During converting compression type, this label is looked up first. If there is no targeting label or the content is lost in the content store, conversion happens.

ktock commented:
Can we move this forward?

sipsma commented:
If there are multiple blobs for a single ref, do each of them count when calculating disk usage of a ref in the prune code? As far as I can see, because multiple blobs per ref are tracked only using leases, that means the cache's metadata still only stores one blob ID per ref (whichever one is created first I suppose). This in turn means that this code will calculate the disk usage using only one of those blobs: https://github.com/moby/buildkit/blob/edc28d1b028bcf240039134dae3176f174e71130/cache/refs.go#L206-L211

sipsma commented on cache/remote.go:
```diff
@@ -104,15 +113,98 @@ func (sr *immutableRef) GetRemote(ctx context.Context, createIfNeeded bool, comp
+		mu.Lock()
+		mprovider.Add(layerP)
+		mu.Unlock()
+		if forceCompression {
```
I feel like this code that creates the blobs with the correct compression type conceptually makes more sense to be in the existing `computeBlobChain` code, somewhere around here maybe: https://github.com/moby/buildkit/blob/edc28d1b028bcf240039134dae3176f174e71130/cache/blobs.go#L57-L58

Not a strong opinion though, so if that's easier said than done I don't think it's a big deal to leave it here.

ktock commented:
@sipsma Thank you for the review.
Moved the blob creation logic into `computeBlobChain` and added a logic to accumulate the size of compression variants.

sipsma commented on cache/refs.go:
```diff
@@ -198,6 +199,16 @@ func (cr *cacheRecord) Size(ctx context.Context) (int64, error) {
+				// accumulate size of compression variant blobs
+				if strings.HasPrefix(k, compressionVariantDigestLabelPrefix) {
+					if cdgst, err := digest.Parse(v); err == nil {
+						if info, err := cr.cm.ContentStore.Info(ctx, cdgst); err == nil {
```
What if you import 2 images that are identical except they use different compression types for their layers and then tried to export each one to their converted compression format. You could then end up in a situation where the blob returned by `getBlob` on this ref also has a compression variant label attached to it, meaning the size would get double counted here, right?

If so, I guess you can just check that `cdgst` doesn't equal `getBlob`? Or something like that.

ktock replied:
Added a check to avoid double count if a label points to the blob itself.

sipsma reviewed (commented):
@ktock commented on one more corner case, otherwise LGTM
